File: golang/producer.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package golang

import (
	"context"
	"fmt"
	"math"
	"sync"
	"time"

	"go.uber.org/atomic"

	"github.com/apache/rocketmq-clients/golang/v5/pkg/utils"
	v2 "github.com/apache/rocketmq-clients/golang/v5/protocol/v2"
	"google.golang.org/protobuf/types/known/durationpb"
)

type Producer interface {
	Send(context.Context, *Message) ([]*SendReceipt, error)
	SendWithTransaction(context.Context, *Message, Transaction) ([]*SendReceipt, error)
	SendAsync(context.Context, *Message, func(context.Context, []*SendReceipt, error))
	BeginTransaction() Transaction
	Start() error
	GracefulStop() error
	isClient
}

type defaultProducer struct {
	po       producerOptions
	pSetting *producerSettings
	cli      *defaultClient

	checker                        *TransactionChecker
	isolated                       sync.Map
	publishingRouteDataResultCache sync.Map
}

func (p *defaultProducer) Start() error {
	err := p.cli.startUp()
	if err == nil {
		return nil
	}
	err2 := p.GracefulStop()
	if err2 != nil {
		return fmt.Errorf("startUp err=%w, shutdown err=%v", err, err2)
	}
	return fmt.Errorf("startUp err=%w", err)
}

var _ = Producer(&defaultProducer{})

func (p *defaultProducer) isClient() {
}

func (p *defaultProducer) isOn() bool {
	return p.cli.on.Load()
}

func (p *defaultProducer) wrapHeartbeatRequest() *v2.HeartbeatRequest {
	return &v2.HeartbeatRequest{
		ClientType: v2.ClientType_PRODUCER,
	}
}

func (p *defaultProducer) takeMessageQueues(plb PublishingLoadBalancer) ([]*v2.MessageQueue, error) {
	return plb.TakeMessageQueues(&p.isolated, p.getRetryMaxAttempts())
}

func (p *defaultProducer) getPublishingTopicRouteResult(ctx context.Context, topic string) (PublishingLoadBalancer, error) {
	item, ok := p.publishingRouteDataResultCache.Load(topic)
	if ok {
		if ret, ok := item.(PublishingLoadBalancer); ok {
			return ret, nil
		}
	}
	route, err := p.cli.getMessageQueues(ctx, topic)
	if err != nil {
		return nil, err
	}
	plb, err := NewPublishingLoadBalancer(route)
	if err != nil {
		return nil, err
	}
	p.publishingRouteDataResultCache.Store(topic, plb)
	return plb, nil
}

func (p *defaultProducer) wrapSendMessageRequest(pMsgs []*PublishingMessage) (*v2.SendMessageRequest, error) {
	smr := &v2.SendMessageRequest{
		Messages: []*v2.Message{},
	}
	for _, pMsg := range pMsgs {
		msgV2, err := pMsg.toProtobuf()
		if err != nil {
			return nil, fmt.Errorf("wrapSendMessageRequest failed, {%v}", err)
		}
		smr.Messages = append(smr.Messages, msgV2)
	}
	return smr, nil
}

var NewProducer = func(config *Config, opts ...ProducerOption) (Producer, error) {
	copyOpt := defaultProducerOptions
	po := &copyOpt
	for _, opt := range opts {
		opt.apply(po)
	}
	cli, err := po.clientFunc(config)
	if err != nil {
		return nil, err
	}
	p := &defaultProducer{
		po:      *po,
		cli:     cli.(*defaultClient),
		checker: po.checker,
	}
	p.cli.initTopics = po.topics
	endpoints, err := utils.ParseTarget(config.Endpoint)
	if err != nil {
		return nil, err
	}
	p.pSetting = &producerSettings{
		clientId:   p.cli.GetClientID(),
		endpoints:  endpoints,
		clientType: v2.ClientType_PRODUCER,
		retryPolicy: &v2.RetryPolicy{
			MaxAttempts: po.maxAttempts,
			Strategy: &v2.RetryPolicy_ExponentialBackoff{
				ExponentialBackoff: &v2.ExponentialBackoff{
					Max:        durationpb.New(time.Duration(0)),
					Initial:    durationpb.New(time.Duration(0)),
					Multiplier: 1,
				},
			},
		},
		requestTimeout:      p.cli.opts.timeout,
		validateMessageType: *atomic.NewBool(true),
		maxBodySizeBytes:    *atomic.NewInt32(4 * 1024 * 1024),
	}
	for _, topic := range po.topics {
		topicResource := &v2.Resource{
			Name:              topic,
			ResourceNamespace: config.NameSpace,
		}
		p.pSetting.topics.Store(topic, topicResource)
	}
	p.cli.settings = p.pSetting
	p.cli.clientImpl = p
	return p, nil
}

func (p *defaultProducer) getRetryMaxAttempts() int {
	return int(p.pSetting.GetRetryPolicy().GetMaxAttempts())
}
func (p *defaultProducer) getNextAttemptDelay(attempt int) time.Duration {
	if attempt <= 0 {
		return time.Duration(0)
	}
	retryPolicy := p.pSetting.GetRetryPolicy().Strategy.(*v2.RetryPolicy_ExponentialBackoff).ExponentialBackoff
	delayNanos := uint32(math.Min(float64(retryPolicy.Initial.AsDuration().Nanoseconds())*math.Pow(float64(retryPolicy.Multiplier), 1.0*float64(attempt-1)), float64(retryPolicy.Max.AsDuration().Nanoseconds())))
	if delayNanos <= 0 {
		return time.Duration(0)
	}
	return time.Duration(delayNanos)
}

func (p *defaultProducer) send1(ctx context.Context, topic string, messageType v2.MessageType,
	candidates []*v2.MessageQueue, pubMessages []*PublishingMessage, attempt int) ([]*SendReceipt, error) {

	ctx = p.cli.Sign(ctx)

	idx := utils.Mod(int32(attempt)-1, len(candidates))
	selectMessageQueue := candidates[idx]

	if p.pSetting.IsValidateMessageType() && !utils.MatchMessageType(selectMessageQueue, messageType) {
		return nil, fmt.Errorf("current message type not match with topic accept message types")
	}

	endpoints := selectMessageQueue.GetBroker().GetEndpoints()

	sendReq, err := p.wrapSendMessageRequest(pubMessages)
	if err != nil {
		return nil, err
	}
	messageCommons := make([]*MessageCommon, 0)
	for _, pubMessage := range pubMessages {
		messageCommons = append(messageCommons, pubMessage.msg.GetMessageCommon())
	}
	p.cli.doBefore(MessageHookPoints_SEND, messageCommons)
	watchTime := time.Now()
	resp, err := p.cli.clientManager.SendMessage(ctx, endpoints, sendReq, p.pSetting.GetRequestTimeout())
	duration := time.Since(watchTime)
	messageHookPointsStatus := MessageHookPointsStatus_OK
	// processSendResponse
	tooManyRequests := false
	if err == nil && resp.GetStatus().GetCode() != v2.Code_OK {
		tooManyRequests = resp.GetStatus().GetCode() == v2.Code_TOO_MANY_REQUESTS
		err = &ErrRpcStatus{
			Code:    int32(resp.Status.GetCode()),
			Message: resp.GetStatus().GetMessage(),
		}
	}
	if err != nil {
		messageHookPointsStatus = MessageHookPointsStatus_ERROR
	}
	p.cli.doAfter(MessageHookPoints_SEND, messageCommons, duration, messageHookPointsStatus)
	maxAttempts := p.getRetryMaxAttempts()
	if err != nil {
		messageIds := make([]string, 0)
		for _, pubMessage := range pubMessages {
			messageIds = append(messageIds, pubMessage.messageId)
		}
		// retry
		for _, address := range endpoints.GetAddresses() {
			p.isolated.Store(utils.ParseAddress(address), true)
		}
		if attempt >= maxAttempts {
			p.cli.log.Errorf("failed to send message(s) finally, run out of attempt times, topic=%s, messageId(s)=%v, maxAttempts=%d, attempt=%d, endpoints=%v, requestId=%s",
				topic, messageIds, maxAttempts, attempt, endpoints, utils.GetRequestID(ctx))
			return nil, err
		}
		// Try to do more attempts.
		nextAttempt := attempt + 1
		// Retry immediately if the request is not throttled.
		if tooManyRequests {
			waitTime := p.getNextAttemptDelay(nextAttempt)
			p.cli.log.Warnf("failed to send message due to too many requests, would attempt to resend after %v, topic=%s, messageId(s)=%v, maxAttempts=%d, attempt=%d, endpoints=%v, requestId=%s",
				waitTime, topic, messageIds, maxAttempts, attempt, endpoints, utils.GetRequestID(ctx))
			time.Sleep(waitTime)
		} else {
			p.cli.log.Warnf("failed to send message, would attempt to resend right now, topic=%s, messageId(s)=%v, maxAttempts=%d, attempt=%d, endpoints=%v, requestId=%s",
				topic, messageIds, maxAttempts, attempt, endpoints, utils.GetRequestID(ctx))
		}
		return p.send1(ctx, topic, messageType, candidates, pubMessages, nextAttempt)
	}

	var res []*SendReceipt
	for i := 0; i < len(resp.GetEntries()); i++ {
		res = append(res, &SendReceipt{
			MessageID:     resp.GetEntries()[i].GetMessageId(),
			TransactionId: resp.GetEntries()[i].GetTransactionId(),
			Offset:        resp.GetEntries()[i].GetOffset(),
			Endpoints:     endpoints,
		})
	}
	if attempt > 1 {
		p.cli.log.Infof("resend message successfully, topic=%s, maxAttempts=%d, attempt=%d, endpoints=%s",
			topic, maxAttempts, attempt, endpoints.String())
	}
	return res, nil
}

func (p *defaultProducer) send0(ctx context.Context, msgs []*UnifiedMessage, txEnabled bool) ([]*SendReceipt, error) {
	// check topic Name
	topicName := msgs[0].GetMessage().Topic
	for _, msg := range msgs {
		if msg.GetMessage().Topic != topicName {
			return nil, fmt.Errorf("messages to send have different topics")
		}
	}

	pubMessages := make([]*PublishingMessage, len(msgs))
	for idx, uMsg := range msgs {
		msg := uMsg.GetMessage()
		var pubMessage *PublishingMessage
		var err error
		pubMessage = uMsg.pubMsg
		if uMsg.pubMsg == nil {
			pubMessage, err = NewPublishingMessage(msg, p.cli.config.NameSpace, p.pSetting, txEnabled)
			if err != nil {
				return nil, err
			}
		}
		pubMessages[idx] = pubMessage
	}

	// check message Type
	messageType := pubMessages[0].messageType
	for _, pubMessage := range pubMessages {
		if pubMessage.messageType != messageType {
			return nil, fmt.Errorf("messages to send have different types, please check")
		}
	}

	var messageGroup *string
	// Message group must be same if message type is FIFO, or no need to proceed.
	if messageType == v2.MessageType_FIFO {
		messageGroup = pubMessages[0].msg.GetMessageGroup()
		for _, pubMessage := range pubMessages {
			if pubMessage.msg.GetMessageGroup() != messageGroup {
				return nil, fmt.Errorf("fifo messages to send have different message groups")
			}
		}
	}
	if _, ok := p.pSetting.topics.Load(topicName); !ok {
		p.pSetting.topics.Store(topicName, &v2.Resource{
			Name:              topicName,
			ResourceNamespace: p.cli.config.NameSpace,
		})
	}
	pubLoadBalancer, err := p.getPublishingTopicRouteResult(ctx, topicName)
	if err != nil {
		return nil, err
	}
	var candidates []*v2.MessageQueue
	if messageGroup == nil {
		candidates, err = p.takeMessageQueues(pubLoadBalancer)
	} else {
		candidates, err = pubLoadBalancer.TakeMessageQueueByMessageGroup(messageGroup)
	}
	if err != nil || len(candidates) == 0 {
		return nil, fmt.Errorf("no broker available to sendMessage")
	}
	return p.send1(ctx, topicName, messageType, candidates, pubMessages, 1)
}

func (p *defaultProducer) Send(ctx context.Context, msg *Message) ([]*SendReceipt, error) {
	if !p.isOn() {
		return nil, fmt.Errorf("producer is not running")
	}
	msgs := []*UnifiedMessage{{
		msg: msg,
	}}
	return p.send0(ctx, msgs, false)
}

func (p *defaultProducer) SendAsync(ctx context.Context, msg *Message, f func(context.Context, []*SendReceipt, error)) {
	if !p.isOn() {
		f(ctx, nil, fmt.Errorf("producer is not running"))
	}
	go func() {
		msgs := []*UnifiedMessage{{
			msg: msg,
		}}
		resp, err := p.send0(ctx, msgs, false)
		f(ctx, resp, err)
	}()
}

func (p *defaultProducer) SendWithTransaction(ctx context.Context, msg *Message, transaction Transaction) ([]*SendReceipt, error) {
	if !p.isOn() {
		return nil, fmt.Errorf("producer is not running")
	}
	t := transaction.(*transactionImpl)
	pubMessage, err := t.tryAddMessage(msg, p.cli.config.NameSpace)
	if err != nil {
		return nil, err
	}
	pubMsgs := []*UnifiedMessage{{
		pubMsg: pubMessage,
	}}
	resp, err := p.send0(ctx, pubMsgs, true)
	if err != nil {
		return nil, err
	}
	err2 := t.tryAddReceipt(pubMessage, resp[0])
	if err2 != nil {
		p.cli.log.Error(err2)
	}
	return resp, err
}

func (p *defaultProducer) GracefulStop() error {
	return p.cli.GracefulStop()
}

func (p *defaultProducer) BeginTransaction() Transaction {
	return NewTransactionImpl(p)
}

func (p *defaultProducer) endTransaction(ctx context.Context, endpoints *v2.Endpoints, messageCommon *MessageCommon,
	messageId string, transactionId string, resolution TransactionResolution) error {

	ctx = p.cli.Sign(ctx)
	request := &v2.EndTransactionRequest{
		Topic: &v2.Resource{
			Name:              messageCommon.topic,
			ResourceNamespace: p.cli.config.NameSpace,
		},
		MessageId:     messageId,
		TransactionId: transactionId,
	}
	switch resolution {
	case COMMIT:
		request.Resolution = v2.TransactionResolution_COMMIT
	case ROLLBACK:
		request.Resolution = v2.TransactionResolution_ROLLBACK
	default:
		request.Resolution = v2.TransactionResolution_TRANSACTION_RESOLUTION_UNSPECIFIED
	}
	requestTimeout := p.pSetting.requestTimeout
	messageHookPoints := MessageHookPoints_COMMIT_TRANSACTION
	if resolution == ROLLBACK {
		messageHookPoints = MessageHookPoints_ROLLBACK_TRANSACTION
	}
	messageCommons := []*MessageCommon{messageCommon}
	p.cli.doBefore(messageHookPoints, messageCommons)
	watchTime := time.Now()
	resp, err := p.cli.clientManager.EndTransaction(ctx, endpoints, request, requestTimeout)
	duration := time.Since(watchTime)
	messageHookPointsStatus := MessageHookPointsStatus_OK
	if err == nil && resp.GetStatus().GetCode() != v2.Code_OK {
		err = &ErrRpcStatus{
			Code:    int32(resp.Status.GetCode()),
			Message: resp.GetStatus().GetMessage(),
		}
	}
	if err != nil {
		messageHookPointsStatus = MessageHookPointsStatus_ERROR
	}
	p.cli.doAfter(messageHookPoints, messageCommons, duration, messageHookPointsStatus)
	return err
}

func (p *defaultProducer) onRecoverOrphanedTransactionCommand(endpoints *v2.Endpoints, command *v2.RecoverOrphanedTransactionCommand) error {
	transactionId := command.GetTransactionId()
	messageId := command.GetMessage().GetSystemProperties().MessageId
	if p.checker == nil {
		return fmt.Errorf("no transaction checker registered, ignore it, messageId=%s, transactionId=%s, endpoints=%v", messageId, transactionId, endpoints)
	}
	messageView := fromProtobuf_MessageView0(command.Message)
	go func(mv *MessageView) {
		resolution := p.po.checker.Check(mv)
		err := p.endTransaction(context.TODO(), endpoints,
			mv.GetMessageCommon(), messageId, transactionId, resolution)
		if err != nil {
			p.cli.log.Errorf("exception raised while ending the transaction, messageId=%s, transactionId=%s, endpoints=%v, err=%w", messageId, transactionId, endpoints, err)
		}
	}(messageView)
	return nil
}

func (p *defaultProducer) onVerifyMessageCommand(endpoints *v2.Endpoints, command *v2.VerifyMessageCommand) error {
	return nil
}

func (p *defaultProducer) SetRequestTimeout(timeout time.Duration) {
	p.cli.opts.timeout = timeout
	p.pSetting.requestTimeout = p.cli.opts.timeout
}

func (p *defaultProducer) IsEndpointUpdated() bool {
	return p.cli.ReceiveReconnect
}

File: php/Producer.php
<?php
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Apache\Rocketmq;

use Apache\Rocketmq\V2\MessagingServiceClient;
use Apache\Rocketmq\V2\Resource;
use Apache\Rocketmq\V2\Message;
use Apache\Rocketmq\V2\SystemProperties;
use Apache\Rocketmq\V2\Settings;
use Apache\Rocketmq\V2\ClientType;
use Apache\Rocketmq\V2\UA;
use Apache\Rocketmq\V2\Language;
use Apache\Rocketmq\V2\TelemetryCommand;
use Apache\Rocketmq\V2\Publishing;

/**
 * Producer — Message producer.
 *
 * Core features: singleton TelemetrySession, PublishingLoadBalancer, retry,
 * transaction support (via TransactionTrait), heartbeat (via HeartbeatManager),
 * interceptor support, Swoole coroutine async support.
 *
 * Send and recall logic is delegated to SendMessageHandler and RecallMessageHandler.
 * Use ProducerBuilder for convenient construction.
 */
class Producer implements TransactionCommitter, ClientTraitProvider
{
    use ClientTrait {
        buildMetadata as public;
        getOperationTimeout as public;
        parseEndpoints as public;
    }
    use TransactionTrait;

    private readonly MessagingServiceClient $client;
    private readonly TelemetrySession $telemetrySession;
    private readonly PublishingRouteManager $routeManager;
    private readonly ProducerSettings $settings;
    protected MessageValidator $validator;
    private bool $isRunning = false;
    private bool $shutdownRequested = false;
    private array $interceptors = [];
    private readonly Logger $logger;
    private readonly HeartbeatManager $heartbeatManager;
    private readonly SendMessageHandler $sendHandler;
    private readonly RecallMessageHandler $recallHandler;

    /**
     * @param string $endpoints gRPC server endpoint
     * @param array $options Configuration options (use ProducerBuilder instead)
     *                      - clientId: string, custom client identifier (default: 'php-producer-{pid}-{time}')
     *                      - maxAttempts: int, max retry attempts on send failure (default: 3)
     *                      - requestTimeout: int, gRPC request timeout in ms (default: 3000)
     *                      - topics: string[], topics to publish messages to (default: [])
     *                      - namespace: string, resource namespace prefix (default: '')
     *                      - credentials: SessionCredentials|null, AK/SK authentication credentials
     *                      - validateMessageType: bool, validate message type against route (default: true)
     *                      - maxBodySizeBytes: int, max message body size in bytes (default: 4194304)
     *                      - tlsCredentials: TlsCredentials|null, TLS/SSL configuration
     *                      - sslEnabled: bool, enable SSL for gRPC channel (default: true)
     * @deprecated Use ProducerBuilder instead for better type safety and IDE support.
     */
    public function __construct(
        private readonly string $endpoints,
        array $options = []
    ) {
        $this->settings = new ProducerSettings($endpoints, $options);
        $this->validator = new MessageValidator(
            $options['maxBodySizeBytes'] ?? 4194304,
            $options['validateMessageType'] ?? true
        );
        $this->logger = Logger::getInstance('Producer');

        $this->client = RpcClientManager::getInstance()->getClient($endpoints, [
            'tlsCredentials' => $this->settings->getTlsCredentials(),
            'sslEnabled' => $this->settings->isSslEnabled(),
        ]);

        $this->telemetrySession = TelemetrySession::getInstance(
            $this->client, $endpoints, $this->settings->getClientId(),
            $this->settings->getCredentials(), $this->settings->getNamespace()
        );
        $this->routeManager = new PublishingRouteManager($this->client, $endpoints, $this);
        $this->heartbeatManager = new HeartbeatManager(
            $this->routeManager, $this->client, $this,
            $this->settings->getTlsCredentials(), $this->settings->isSslEnabled()
        );

        $metadataBuilder = fn(?int $timeoutMs = null) => $this->buildMetadata($timeoutMs);
        $callOptionsResolver = fn(?int $overrideTimeout = null) => $this->getCallOptions($overrideTimeout);
        $operationTimeoutFn = fn(string $op) => $this->getOperationTimeout($op);
        $interceptorExecutor = function (string $hookPoint, array $context = []) {
            $this->executeInterceptors($hookPoint, $context);
        };

        $this->sendHandler = new SendMessageHandler(
            $this->client,
            $this->settings,
            $this->validator,
            $this->routeManager,
            $interceptorExecutor,
            $metadataBuilder,
            $callOptionsResolver,
            $operationTimeoutFn,
        );

        $this->recallHandler = new RecallMessageHandler(
            $this->client,
            $this->settings,
            $metadataBuilder,
            $callOptionsResolver,
        );
    }

    // ==================== Lifecycle ====================

    public function start(): void
    {
        if ($this->isRunning) {
            return;
        }

        try {
            Logger::getInstance('Producer')->info("Begin to start the rocketmq producer, clientId={$this->settings->getClientId()}");
            $this->establishTelemetrySession();
            $this->registerSettingsCallback();
            $this->registerTransactionCheckerCallback();

            $this->routeManager->warmUp($this->settings->getTopics());

            $this->isRunning = true;
            $this->heartbeatManager->start();

            Logger::getInstance('Producer')->info("The rocketmq producer starts successfully, clientId={$this->settings->getClientId()}");
        } catch (\Exception $e) {
            Logger::getInstance('Producer')->error("Failed to start: " . $e->getMessage());
            $this->shutdown();
            throw $e;
        }
    }

    public function shutdown(): void
    {
        if (!$this->isRunning) {
            return;
        }

        $this->shutdownRequested = true;
        $this->logger->info("Begin to shutdown the rocketmq producer, clientId={$this->settings->getClientId()}");

        if (SwooleCompat::isAvailable() && SwooleCompat::inCoroutine()) {
            \Swoole\Coroutine::sleep(1);
        }

        $this->heartbeatManager->stop();
        $this->heartbeatManager->notifyClientTermination();

        if ($this->telemetrySession) {
            $this->telemetrySession->close();
        }

        $this->isRunning = false;
        $this->logger->info("Shutdown the rocketmq producer successfully, clientId={$this->settings->getClientId()}");
    }

    public function __destruct()
    {
        $this->shutdown();
    }

    // ==================== Send ====================

    /**
     * Send a message
     * @param Message $message to send
     * @return array Send result containing:
     * - messageId: messageId
     * - messageQueue: messageQueue
     * - offset: offset
     * - requestId: requestId
     * - sendTime: sendTime
     * - transactionId: transactionId
     * - transactionState: transactionState
     * @throws \Exception if producer is not running
     */
    public function send(Message $message): array
    {
        if (!$this->isRunning) {
            throw new \RuntimeException("Producer is not running now");
        }
        return $this->sendHandler->send($message);
    }

    /**
     * Send a message asynchronously
     * @param Message $message to send
     * @return \Generator|mixed|null | void
     */
    public function sendAsync(Message $message): array|\Generator
    {
        if (!$this->isRunning) {
            throw new \RuntimeException("Producer is not running now");
        }
        return $this->sendHandler->sendAsync($message);
    }

    // ==================== Batch Send ====================

    /**
     * Send a batch of messages
     * @param array $messages to send
     * @return array Send result containing:
     * @throws \Exception if producer is not running
     */
    public function sendBatch(array $messages): array
    {
        if (!$this->isRunning) {
            throw new \RuntimeException("Producer is not running now");
        }
        return $this->sendHandler->sendBatch($messages);
    }

    /**
     * Send a batch of messages asynchronously
     * @param array $messages to send
     * @return \Generator|mixed|null | void
     */
    public function sendBatchAsync(array $messages): array|\Generator
    {
        if (!$this->isRunning) {
            throw new \RuntimeException("Producer is not running now");
        }
        return $this->sendHandler->sendBatchAsync($messages);
    }

    // ==================== Convenience Send Methods ====================

    /**
     * Send a priority message
     * @param $topic  string Topic name
     * @param $body  string Message body
     * @param $priority  int Message priority
     * @param $tag  string Message tag
     * @return array Send result containing:
     * @throws \Exception if producer is not running
     */
    public function sendPriorityMessage(string $topic, string $body, int $priority, string $tag = ''): array
    {
        if (!$this->isRunning) {
            throw new \RuntimeException("Producer is not running now");
        }
        return $this->send($this->sendHandler->buildConvenienceMessage($topic, $body, $tag, function (SystemProperties $sp) use ($priority) {
            $sp->setPriority($priority);
        }));
    }

    /**
     * Send a delayed message
     * @param $topic string Topic name
     * @param $body string Message body
     * @param $deliveryTimestampUnixSec Unix timestamp
     * @param $tag  string Message tag
     * @return array Send result containing:
     * @throws \Exception if producer is not running
     */
    public function sendDelayedMessage(string $topic, string $body, int $deliveryTimestampUnixSec, string $tag = ''): array
    {
        if (!$this->isRunning) {
            throw new \RuntimeException("Producer is not running now");
        }
        $ts = new \Google\Protobuf\Timestamp();
        $ts->setSeconds($deliveryTimestampUnixSec);
        $ts->setNanos(0);

        return $this->send($this->sendHandler->buildConvenienceMessage($topic, $body, $tag, function (SystemProperties $sp) use ($ts) {
            $sp->setDeliveryTimestamp($ts);
        }));
    }

    /**
     * Send a FIFO message
     *
     * @param $topic string Topic name
     * @param $body string Message body
     * @param $messageGroup  string FIFO message group
     * @param $tag  string  Message tag
     * @return array Send result containing:
     * @throws \Exception if producer is not running
     */
    public function sendFifoMessage(string $topic, string $body, string $messageGroup, string $tag = ''): array
    {
        if (!$this->isRunning) {
            throw new \RuntimeException("Producer is not running now");
        }
        return $this->send($this->sendHandler->buildConvenienceMessage($topic, $body, $tag, function (SystemProperties $sp) use ($messageGroup) {
            $sp->setMessageGroup($messageGroup);
        }));
    }

    // ==================== Recall ====================

    /**
     * Recall a message
     * @param $topic string Topic name
     * @param $recallHandle recall handle
     * @return array recall result containing:
     */
    public function recallMessage(string $topic, string $recallHandle): array
    {
        if (!$this->isRunning) {
            throw new \RuntimeException("Producer is not running now");
        }
        return $this->recallHandler->recall($topic, $recallHandle);
    }

    public function recallMessageAsync(string $topic, string $recallHandle): array|\Generator
    {
        if (!$this->isRunning) {
            throw new \RuntimeException("Producer is not running now");
        }
        return $this->recallHandler->recallAsync($topic, $recallHandle);
    }

    // ==================== Interceptors ====================

    /**
     * Add an interceptor
     * @param MessageInterceptor $interceptor The interceptor to add
     * @return $this  For method chaining
     * @throws \Exception if producer is not running
     */
    public function addInterceptor(MessageInterceptor $interceptor): self
    {
        $this->interceptors[] = $interceptor;
        return $this;
    }

    /**
     * Execute interceptors
     * @param $hookPoint
     * @param $context
     * @return void
     */
    public function executeInterceptors(string $hookPoint, array $context = []): void
    {
        if (empty($this->interceptors)) {
            return;
        }
        foreach ($this->interceptors as $interceptor) {
            try {
                $interceptor->intercept($hookPoint, $context);
            } catch (\Exception $e) {
                $this->logger->warning("Interceptor failed at {$hookPoint}: " . $e->getMessage());
            }
        }
    }

    // ==================== Getters ====================

    public function getClientId(): string { return $this->settings->getClientId(); }
    public function isRunning(): bool { return $this->isRunning; }

    // ==================== ClientTrait Required Methods ====================

    protected function getCredentials(): ?SessionCredentials { return $this->settings->getCredentials(); }
    protected function getClientIdValue(): string { return $this->settings->getClientId(); }
    protected function getNamespaceValue(): string { return $this->settings->getNamespace(); }

    // ==================== Private: Telemetry & Settings ====================

    private function establishTelemetrySession(): void
    {
        $ua = new UA();
        $ua->setLanguage(Language::PHP);
        $ua->setVersion(ClientConstants::CLIENT_VERSION);

        $publishing = new Publishing();
        $topicResources = [];
        foreach ($this->settings->getTopics() as $topicName) {
            $topicResource = new Resource();
            $topicResource->setName($topicName);
            $topicResources[] = $topicResource;
        }
        $publishing->setTopics($topicResources);

        $settings = new Settings();
        $settings->setClientType(ClientType::PRODUCER);
        $settings->setUserAgent($ua);
        $settings->setPublishing($publishing);

        $command = new TelemetryCommand();
        $command->setSettings($settings);

        if (!$this->telemetrySession->syncSettings($command)) {
            throw new \RuntimeException("Failed to establish Telemetry Session");
        }
        SwooleCompat::sleep(500000);
    }

    private function registerSettingsCallback(): void
    {
        $self = $this;
        $this->telemetrySession->setOnSettingsChange(function ($settings) use ($self) {
            $self->onServerSettings($settings);
        });
    }

    private function onServerSettings(object $settings): void
    {
        $this->logger->info("Processing server settings");

        if ($settings->hasPublishing()) {
            $publishing = $settings->getPublishing();
            if ($publishing->getMaxBodySize() > 0) {
                $this->validator->setMaxBodySizeBytes($publishing->getMaxBodySize());
                $this->logger->info("Updated maxBodySize from server: {$this->validator->getMaxBodySizeBytes()}");
            }
            $this->validator->setValidateMessageType($publishing->getValidateMessageType());
            $this->logger->info("Updated validateMessageType from server: " . ($this->validator->isValidateMessageType() ? 'true' : 'false'));
        }

        $this->settings->applyServerBackoffPolicy($settings, $this->logger);
    }

    // ==================== TransactionTrait Delegation ====================
    // These methods provide the interface that TransactionTrait requires.
    // They delegate to private properties or SendMessageHandler.

    private function validateMessage(Message $message): void
    {
        $this->validator->validateMessage($message);
    }

    private function detectMessageType(Message $msg, bool $txEnabled = false): int
    {
        return $this->sendHandler->detectMessageType($msg, $txEnabled);
    }

    private function wrapTransactionMessageRequest(array $messages, object $messageQueue): V2\SendMessageRequest
    {
        return $this->sendHandler->wrapTransactionMessageRequest($messages, $messageQueue);
    }

    private function sendMessageWithRetry(V2\SendMessageRequest $request, Message $message, array $candidates, int $maxAttempts): array
    {
        return $this->sendHandler->sendMessageWithRetry($request, $message, $candidates, $maxAttempts);
    }

    // ==================== TransactionTrait Infrastructure Delegation ====================
    // Protected methods so the trait can access infrastructure without touching
    // private properties. Override in test fakes for isolation.

    protected function getPublishingLoadBalancer(string $topic): object
    {
        return $this->routeManager->getPublishingLoadBalancer($topic);
    }

    protected function getIsolatedBrokerNames(): array
    {
        return $this->routeManager->getIsolatedBrokerNames();
    }

    protected function getSettingsMaxAttempts(): int
    {
        return $this->settings->getMaxAttempts();
    }

    protected function getSettingsTlsCredentials(): ?TlsCredentials
    {
        return $this->settings->getTlsCredentials();
    }

    protected function isSettingsSslEnabled(): bool
    {
        return $this->settings->isSslEnabled();
    }

    protected function getClientForRpc(): MessagingServiceClient
    {
        return $this->client;
    }

    protected function getTelemetrySession(): TelemetrySession
    {
        return $this->telemetrySession;
    }

    protected function getLogger(): Logger
    {
        return $this->logger;
    }
}
